Explanation and examples added for various QuantumCircuit methods Fixes #3400

sumitpuri opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [X] I have added the tests to cover my changes.
- [X] I have updated the documentation accordingly.
- [X] I have read the CONTRIBUTING document.
-->

### Summary

Fixes #3400 by adding explanations for various QuantumCircuit methods.

### Details and comments

The gates (ccx, ch, crz, cswap, cu1, cu3, cx, cy, cz, h, iden, rx, ry, rz, s, sdg, swap, t, tdg, u1, u2, u3, x, y, z) now have an example and explanation of what the respective gate does.

nonhermitian commented:
The examples should be in jupyter-execute blocks so as they are run at docs build time. This way errors in the docs are caught.

sumitpuri commented:
@nonhermitian So are you suggesting to split the explanation and the example circuit? Are the one line explanations okay in these files? Can you please share a sample screenshot of where the examples need to be added?

sumitpuri commented:
I was doing a PR for another bug #3684 but seems like it got tied to this PR instead.

kdk commented:
The explanations and examples can stay together in the same docstrings, but the example should be prefaced with `.. jupyter-execute::` so that they will be run and displayed in the auto-generated documentation.  For example, https://github.com/Qiskit/qiskit-terra/blob/7718e3576d0c9d77991c8c4b813ac05b7b1aa940/qiskit/circuit/quantumcircuit.py#L76
builds to the Examples section on https://qiskit.org/documentation/api/qiskit.circuit.QuantumCircuit.html#quantumcircuit .

kdk commented:
> I was doing a PR for another bug #3684 but seems like it got tied to this PR instead.

You can revert that commit from this PR by checking out the `fixissue3640branch` branch and running `git revert b8b010e`.

sumitpuri commented:
Thank you @kdk Can you help me fix the two checks that are failing.

kdk commented on qiskit/extensions/standard/crz.py:
```diff
@@ -57,7 +57,27 @@ def inverse(self):
+        Construct a circuit with cRz gate.
+
+        .. jupyter-execute::
+
```
```suggestion

import numpy
```

sumitpuri commented:
Thank you @kdk , could you plz also guide how did you catch the error? Where are we supposed to look as make lint worked fine for me and it didn't catch this. Also the failed checks did not show this missing import error anywhere.

nonhermitian commented:
So I am still not quite onboard with these docs. The point should be to explain what the function is doing. In this case we are just appending a gate to a circuit. The docstrings should say what the gate does, give the circuit representation and perhaps the matrix definition used in qiskit, eg think the someone non canonical cx. The state vector stuff can go as It is out of the functions scope and is not executed by jupyter.

sumitpuri commented:
Sure @nonhermitian Can you please check the below text for X gate and confirm if it meets your expectations? I will then do the similar for other gates as well.

**X Gate**

    """Apply X gate to a specified qubit (q).
    The X gate has the property of flipping the |0⟩ state to |1⟩, and vice versa. It corresponds to a pi rotation about the x-axis on the Bloch sphere.

    Circuit Representation:
	        ┌───┐
	q_0: |0>┤ X ├
	        └───┘
    Matrix Representation:
        [ 0 ,
          1 ]
	"""

nonhermitian commented:
I would have something like this in a `jupyter-execute` block:
```python
qc = QuantumCircuit(1)
qc.x(0)
qc.draw()
```
as well as this:
```python
from qiskit.extensions.standard.x import XGate
XGate().to_matrix()
```
along with the appropriate text that says what these items are.  That way the resulting circuit with the gate and the matrix definition are all computed at doc build time.  The matrix definition is not completely appropriate here given that it is part of a different class, but accessing gates via their inclusion in a quantum circuit is the entry point for most users, so it is important that we define as much as possible here.

sumitpuri commented:
@kdk How can we debug why checks are failing? The line numbers marked by checks do not give anything in the azure pipeline. 

![image](https://user-images.githubusercontent.com/6925028/72065864-d6eb6d80-32a4-11ea-8528-374868918e02.png)

nonhermitian commented:
This looks to be an unrelated issue with the `sphinx-tabs` package used in the docs:
```
Running setup.py install for sphinx-tabs: finished with status 'error'
    ERROR: Command errored out with exit status 1:
     command: /Users/runner/hostedtoolcache/Python/3.8.0/x64/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-qf2hpjhh/sphinx-tabs/setup.py'"'"'; __file__='"'"'/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-qf2hpjhh/sphinx-tabs/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-record-dxnb_16o/install-record.txt --single-version-externally-managed --compile
         cwd: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-qf2hpjhh/sphinx-tabs/
    Complete output (10 lines):
    running install
    running build
    running build_py
    package init file 'sphinx_tabs/__init__.py' not found (or not a regular file)
    package init file 'sphinx_tabs/test/__init__.py' not found (or not a regular file)
    running install_lib
    warning: install_lib: 'build/lib' does not exist -- no Python modules to install
    
    running install_data
    error: can't copy 'LICENSE': doesn't exist or not a regular file

```

sumitpuri commented:
Got it, @nonhermitian so will this suffice - It has the info on what the gate does at the top followed by Circuit Representation and Matrix Representation?

    """Apply X gate to a specified qubit (q).
    The X gate has the property of flipping the |0⟩ state to |1⟩, and vice versa.
    It corresponds to a pi rotation about the x-axis on the Bloch sphere.

    Examples:

        Circuit Representation:

        .. jupyter-execute::

            from qiskit import QuantumCircuit

            circuit = QuantumCircuit(1)
            circuit.x(0)
            circuit.draw()

        Matrix Representation:

        .. jupyter-execute::

            from qiskit.extensions.standard.x import XGate
            XGate().to_matrix()
    """

nonhermitian commented:
A minor correction would be something like:
```rst
An X-gate implements a pi rotation about the x-axis of the Bloch sphere.

This gate is canonically used to rotate the qubit state from |0⟩ to |1⟩, or vise versa.  
```
It is an open question as to whether or not to write the math expressions using rst math so that the docs renders them in LaTeX.  This reduces readability for the text version, so others like @kdk and @aasfaw may have thoughts on this.

kdk commented:
> @kdk How can we debug why checks are failing? The line numbers marked by checks do not give anything in the azure pipeline.
> 

The error message isn't available from azure through github, unfortunately. You have to click "View more details on Azure Pipelines" and follow through to the failing stage to find the error message. In this case https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=6914&view=logs&j=8e751551-5bd1-5d1b-509f-24f862c2e768&t=ab63e2c3-1c4d-5327-2f64-6cf6baf5d891&l=335 .

You can also run `tox -edocs` to test building the documentation locally.

sumitpuri commented:
> > @kdk How can we debug why checks are failing? The line numbers marked by checks do not give anything in the azure pipeline.
> 
> The error message isn't available from azure through github, unfortunately. You have to click "View more details on Azure Pipelines" and follow through to the failing stage to find the error message. In this case https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=6914&view=logs&j=8e751551-5bd1-5d1b-509f-24f862c2e768&t=ab63e2c3-1c4d-5327-2f64-6cf6baf5d891&l=335 .
> 
> You can also run `tox -edocs` to test building the documentation locally.

Thanks @kdk so how do these errors get fixed. e.g.  
error: can't copy 'LICENSE': doesn't exist or not a regular file

sumitpuri commented:
@nonhermitian I have incorporated your comments. Here is the summary of what all I have added. 

File | Description (rotation)| Description (Canonical) | Circuit Representation | Matrix Representation
---|---|---|---|---
ccx.py | - | Yes | Yes | Yes
ch.py | - | Yes | Yes | Yes
crz.py | Yes | - | Yes | -
cswap.py | - | Yes | Yes | -
cu1.py | Yes | - | Yes | -
cu3.py | Yes | - | Yes | -
cx.py | Yes | - | Yes | Yes
cy.py | Yes | - | Yes | -
cz.py | Yes | - | Yes | Yes
h.py | Yes | Yes | Yes | Yes
iden.py | - | Yes | Yes | Yes
rx.py | Yes | - | Yes | Yes
ry.py | Yes | - | Yes | Yes
rz.py | Yes | - | Yes | Yes
s.py | Yes | - | Yes | Yes
swap.py | - | Yes | Yes | Yes
t.py | Yes | - | Yes | Yes
u1.py | Yes | - | Yes | Yes
u2.py | Yes | - | Yes | Yes
u3.py | Yes | - | Yes | Yes
x.py | Yes | Yes | Yes | Yes
y.py | Yes | Yes | Yes | Yes
z.py | Yes | Yes | Yes | Yes

The files that I have not worked that were present in the folder are below. Let me know if these files would need the same as above.

| File |
|---|
barrier.py
crx.py
cry.py
ms.py
multi_control_rotation_gates.py
multi_control_toffoli_gate.py
multi_control_u1_gate.py
r.py
relative_phase_toffoli.py
rxx.py
rzz.py

kdk commented on qiskit/extensions/standard/rx.py:
```diff
@@ -59,7 +59,31 @@ def to_matrix(self):
+        .. jupyter-execute::
+
+            from qiskit.extensions.standard.rx import RXGate
+            RXGate().to_matrix()
```
From https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=6971&view=logs&j=23e88340-9b01-571b-4128-239215012d30&t=fd5b7ae3-9f70-5b67-b24e-b010b8098eb2&l=273

```suggestion
            RXGate(numpy.pi/2).to_matrix()
```

sumitpuri commented:
@kdk Did you get a chance to take a look into this PR?

sumitpuri commented on qiskit/extensions/standard/u3.py:
```diff
@@ -52,7 +52,33 @@ def to_matrix(self):
+            circuit.u3(theta,phi,lam,0)
+            circuit.draw()
+
+        Matrix Representation:
```
Done, removed Matrix Representation from U1, U2, U3 gates.

sumitpuri commented on qiskit/extensions/standard/cswap.py:
```diff
@@ -57,7 +57,23 @@ def inverse(self):
+            circuit = QuantumCircuit(3)
+            circuit.cswap(0,1,2)
+            circuit.draw()
+    """
```
Added matrix representation for cswap gate as well.

sumitpuri commented on qiskit/extensions/standard/h.py:
```diff
@@ -55,7 +55,30 @@ def to_matrix(self):
+    """Apply Hadamard (H) gate to a specified qubit (q).
+    An H-gate implements a combination of rotations of the qubit state vector - pi rotation about
+    z-axis followed by pi/2 rotation about the y-axis of the Bloch sphere.
+    This gate is canonically used to rotate the qubit state from |0⟩ to |+⟩ or |1⟩ to |-⟩.
```
Updated this description.

sumitpuri commented on qiskit/extensions/standard/u3.py:
```diff
@@ -52,7 +52,33 @@ def to_matrix(self):
-    """Apply u3 to q."""
+    """Apply U3 gate with angle theta, phi, and lam to a specified qubit (q).
+    A U3 gate implements a theta, phi, and lam radian rotation of the qubit state vector about
+    the x, y, and z-axis, respectively of the Bloch sphere.
```
Done

sumitpuri commented on qiskit/extensions/standard/u2.py:
```diff
@@ -59,7 +59,32 @@ def to_matrix(self):
-    """Apply u2 to q."""
+    """Apply U2 gate with angle phi and lam to a specified qubit (q).
+    A U2-gate implements a phi and lam radian rotation of the qubit state vector about the
+    x and z-axis, respectively of the Bloch sphere.
```
Done

sumitpuri commented on qiskit/extensions/standard/u1.py:
```diff
@@ -51,7 +51,31 @@ def to_matrix(self):
-    """Apply u1 with angle theta to q."""
+    """Apply U1 gate with angle theta to a specified qubit (q).
+    A U1-gate implements a theta radian rotation of the qubit state vector about the
+    z-axis of the Bloch sphere.
```
Done

kdk commented on qiskit/extensions/standard/cswap.py:
```diff
@@ -55,12 +56,23 @@ def inverse(self):
 
+    def to_matrix(self):
+        """Return a Numpy.array for the Fredkin (CSWAP) gate."""
+        return numpy.array([[1, 0, 0, 0, 0, 0, 0, 0],
```
Might need to be aligned to qiskit's qubit indexing convention (I think this is causing the test failures):
```
>>> qk.quantum_info.Operator(FredkinGate()).data.real
array([[1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1.]])
```

sumitpuri replied:
done

kdk commented on qiskit/extensions/standard/cu3.py:
```diff
@@ -65,22 +65,21 @@ def inverse(self):
-    A cU3-gate implements a theta, phi, and lam radian rotation of the qubit state vector about the
-    x, y, and z-axis, respectively of the Bloch sphere when the control qubit is in state |1>.
+    A cU3 gate implements a theta, phi, and lam radian rotation of the qubit state vector about the
+    x, y, and z axis, respectively of the Bloch sphere when the control qubit is in state |1>.
```
Should be updated to match definition of U3 (or just say 'implements a U3(theta,phi,lam) on target if control...')

sumitpuri replied:
done

kdk commented on qiskit/extensions/standard/rz.py:
```diff
@@ -48,25 +49,38 @@ def inverse(self):
         """
         return RZGate(-self.params[0])
 
+    def to_matrix(self):
```
For now, to_matrix on RZGate should be left out (because it differs by a global phase from its definition in terms of U1, which is causing some test failures). This problem is fixed by #3472 so we can add the to_matrix (and its definition in the docs) there.

kdk replied:
Or if you wanted to add the parameterized matrix definition to the docs statically, you could take it from that PR:

```
    .. math::

        U_{\text{RZ}}(\theta)
            = \exp\left(-i \frac{\theta}{2} \sigma_Z \right)
            = \begin{bmatrix}
                e^{-i \theta/2} & 0 \\
                0 & e^{i \theta/2}
            \end{bmatrix}
```

sumitpuri replied:
Removed the to_matrix.

sumitpuri commented:
@kdk I have incorporated the comments, the tests are still failing though.

kdk commented:
> @kdk I have incorporated the comments, the tests are still failing though.

#3631 moved the definitions of some of the controlled gate classes into the files of their base gates (e.g. cx and ccx are now defined in x.py). Normally changes on master are automatically merged in to PR branches, but that change will have to be resolved manually.

sumitpuri commented:
ok @kdk so what is the suggested action item for me?

kdk commented on qiskit/extensions/standard/swap.py:
```diff
@@ -112,9 +133,43 @@ def inverse(self):
+        return numpy.array([[1, 0, 0, 0, 0, 0, 0, 0],
+                            [0, 1, 0, 0, 0, 0, 0, 0],
+                            [0, 0, 1, 0, 0, 0, 0, 0],
+                            [0, 0, 0, 0, 0, 1, 0, 1],
```
```suggestion
                            [0, 0, 0, 0, 0, 1, 0, 0],
```

sumitpuri commented:
Thanks @kdk However, it is failing here:
But I don't see anything wrong in the code.

![image](https://user-images.githubusercontent.com/6925028/73687702-40864e00-4690-11ea-820a-9dfabdba039a.png)

kdk commented on qiskit/extensions/standard/swap.py:
```diff
@@ -112,9 +133,43 @@ def inverse(self):
+
+        .. jupyter-execute::
+
+            from qiskit.extensions.standard.cswap import FredkinGate
```
```suggestion
            from qiskit.extensions.standard.swap import FredkinGate
```

kdk commented on qiskit/extensions/standard/x.py:
```diff
@@ -185,7 +233,29 @@ def to_matrix(self):
+
+        .. jupyter-execute::
+
+            from qiskit.extensions.standard.ccx import ToffoliGate
```
```suggestion
            from qiskit.extensions.standard.x import ToffoliGate
```

sumitpuri commented on qiskit/extensions/standard/z.py:
```diff
@@ -112,7 +136,32 @@ def to_matrix(self):
+
+        .. jupyter-execute::
+
+            from qiskit.extensions.standard.cz import CzGate
```
done

sumitpuri commented:
Thanks @kdk all tests have passed this time.

sumitpuri commented:
Hi @kdk 

In rz.py, there's just a to_matrix definition, not the .. jupyter-execute::

Are you suggesting to remove def to_matrix(self) code?

![image](https://user-images.githubusercontent.com/6925028/73905215-2ef59f80-4864-11ea-8fd0-dd0f658b2aae.png)

kdk commented:
> Are you suggesting to remove def to_matrix(self) code?
Yes, it had caused a test failure earlier due to a phase difference between Rz and U1 (not sure why the test currently passes) https://github.com/Qiskit/qiskit-terra/pull/3683#discussion_r373145628

There are some additional merge conflicts, I'll take care of removing it there.
